java/arcs/core/host/ArcState.kt

```diff
@@ -51,7 +51,12 @@ enum class ArcState {
      * [Deleted] implies [Stopped], but furthermore, subsequent attempts to restart this
      * [Arc] will fail, and potentially any data used by the [Arc] may be reclaimed.
      */
-    Deleted
+    Deleted;
+
+    /**
+     * When an [Arc] moves to the [Error] state, the cause may be attached here.
+     */
+    var errorCause: Exception? = null
```

Comment:
Would it make sense to convert these `enum`s to `sealed class`es?

Reply:
I tried that yesterday; sealed classes are actually fairly awkward given how the enum is used around the code. I'm finishing up a change to using a data class instead.

Reply:
data classes can be sealed classes BTW, e.g. 

```
sealed class ArcState {
   object NotStarted : ArcState
   object Running : ArcState 
   data class Error(args) : ArcState { ... }
}
```

sealed class just prevents not handling all cases in a when{} and can limit extension.

Reply:
I tried that, it doesn't fit especially well with the usage. I think the data class approach behaves more like an enum from a usage POV.

- With a sealed class, implicit typing changes: `val state = ArcState.Running` ends up typed as `ArcState.Running`, not `ArcState`. This pops up in quite a few places and has follow-on effects in other variables decls as well. It's generally more convenient to have a single class with multiple (mostly static) instances, rather than a parent class with multiple sub-classes.
- The Error class ends up being a bit weird, with many equality checks having to change to `is`, and comparisons either moving from `x == ArcState.Error` to `x == ArcState.Error()`, or introducing a static overload on the `Error` name, both of which have issues.

The data class approach doesn't have a problem with extension (data classes are final and I've made the constructor private, so users can only access the static instances and factory methods), but you're right that it doesn't have the all-cases-in-when detection.

Reply:
Fair enough, thanks for the explanation, it helped with the review so I know why you chose this design. :)